### PR TITLE
Add support for pattern expressions via subqueries

### PR DESCRIPTION
**Pull request by s1ck**

### CODE REVIEW

File: caps-core/src/main/scala/org/opencypher/caps/api/expr/Expr.scala [hunk 1]
```diff
@@ -354,8 +353,8 @@ final case class FalseLit() extends BoolLit(false)()
 
 // Pattern Predicate Expression
 
-final case class PatternExpr(predicateField: Var, pattern: Pattern[Expr])(val cypherType: CypherType = CTBoolean) extends Expr {
+final case class ExistsPatternExpr(predicateField: Var, ir: CypherQuery[Expr])(val cypherType: CypherType = CTBoolean) extends Expr {
```
Comment (Mats-SX) [inline caps-core/src/main/scala/org/opencypher/caps/api/expr/Expr.scala]: Hm, will this embedding of the IR into the expression not cause layering issues for subsequent phases?

Comment (Mats-SX) [reply caps-core/src/main/scala/org/opencypher/caps/api/expr/Expr.scala]: I mean, the IR is consumed by the LogicalPlanner; but now (some of) it is also propagated?

File: caps-core/src/main/scala/org/opencypher/caps/impl/logical/LogicalPlanner.scala [hunk 2]
```diff
@@ -177,41 +177,53 @@ class LogicalPlanner(producer: LogicalOperatorProducer)
...
         producer.planFilter(v, acc)
-      case (acc, pe: PatternExpr) =>
-        val patternPlan = planComponentPattern(acc, pe.pattern, context.sourceGraph)
-        val withPredicate = producer.planPatternPredicate(pe, acc, patternPlan)
-        producer.planFilter(pe, withPredicate)
+
+      case (acc, ex: ExistsPatternExpr) =>
+        val innerPlan = this(ex.ir)
+        val Predicate = producer.planExistsPatternPredicate(ex, acc, innerPlan)
+        producer.planFilter(ex, Predicate)
```
Comment (Mats-SX) [inline caps-core/src/main/scala/org/opencypher/caps/impl/logical/LogicalPlanner.scala]: Should use lower-case initial character, otherwise this looks like an object.

File: caps-core/src/main/scala/org/opencypher/caps/api/expr/Expr.scala (hunk 1 above)
Comment (Mats-SX) [reply caps-core/src/main/scala/org/opencypher/caps/api/expr/Expr.scala]: OTOH, it does heavily signal the subquery aspects of this feature, which is appropriate.

File: caps-core/src/main/scala/org/opencypher/caps/impl/logical/LogicalPlanner.scala [hunk 3]
```diff
@@ -222,23 +234,33 @@ class LogicalPlanner(producer: LogicalOperatorProducer)
...
       case func: FunctionExpr =>
         val projectArg = planInnerExpr(func.expr, in)
         producer.projectExpr(func, projectArg)
-      case pe : PatternExpr =>
-        val patternPlan = planComponentPattern(in, pe.pattern, context.sourceGraph)
-        producer.planPatternPredicate(pe, in, patternPlan)
+
+      case ex: ExistsPatternExpr =>
+        val innerPlan = this(ex.ir)
+        producer.planExistsPatternPredicate(ex, in, innerPlan)
```
Comment (Mats-SX) [inline caps-core/src/main/scala/org/opencypher/caps/impl/logical/LogicalPlanner.scala]: Should we allow pattern predicates as inner expressions?

Comment (Mats-SX) [reply caps-core/src/main/scala/org/opencypher/caps/impl/logical/LogicalPlanner.scala]: Ah, for `AND`, `OR` etc of course, although this permits it for any expression position.

File: caps-core/src/main/scala/org/opencypher/caps/impl/parse/rewriter/extractSubqueryFromPatternExpression.scala [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+    /**
+      * WHERE (a)-[:R]->({foo:true})-->()... AND a.age > 20
+      *
+      * to
+      *
+      * WHERE EXISTS (
+      *   MATCH (a)-[e0]->(v0)-[e1]->(v1)...
+      *   WHERE e0:R AND v0.foo = true
+      *   RETURN a, true
+      * ) AND a.age > 20
```
Comment (Mats-SX) [inline caps-core/src/main/scala/org/opencypher/caps/impl/parse/rewriter/extractSubqueryFromPatternExpression.scala]: I think the parentheses should be `{` and `}` here.

File: caps-core/src/main/scala/org/opencypher/caps/impl/util/MapUtils.scala [hunk 5]
```diff
@@ -25,6 +25,18 @@ trait MapUtils {
         case (k, Some(v1), _) => k -> v1
         case (k, _, Some(v2)) => k -> v2
       }.toMap
+
+  def submap[A, B](keys: Seq[A], a: Map[A, B]): Map[A, B] = {
+    val filteredKeys = keys.filter(a.contains)
+    filteredKeys.zip(filteredKeys.map(a)).toMap
```
Comment (Mats-SX) [inline caps-core/src/main/scala/org/opencypher/caps/impl/util/MapUtils.scala]: Isn't this just `a.filterKeys(keys.contains)`?

File: caps-core/src/test/scala/org/opencypher/caps/impl/typer/SchemaTyperTest.scala [hunk 6]
```diff
@@ -325,10 +325,6 @@ class SchemaTyperTest extends BaseTestSuite with Neo4jAstTestSupport with Mockit
     // assertExpr.from("percentileDisc([1, 3.14][$param], 3.14)") shouldHaveInferredType CTInteger
   }
 
-  test("typing of pattern expression") {
-    assertExpr.from("(a)-[:KNOWS]->(b)") shouldHaveInferredType CTBoolean
-  }
```
Comment (Mats-SX) [inline caps-core/src/test/scala/org/opencypher/caps/impl/typer/SchemaTyperTest.scala]: Why was this deleted?

File: caps-spark/src/main/scala/org/opencypher/caps/impl/spark/SparkSQLExprMapper.scala [hunk 7]
```diff
@@ -71,6 +71,7 @@ object SparkSQLExprMapper {
       * @param expr    the expression to be evaluated.
       * @param df      the dataframe containing the data over which the expression should be evaluated.
       * @param context context with helper functions, such as column names.
+      * @param context context with helper functions, such as column names.
```
Comment (Mats-SX) [inline caps-spark/src/main/scala/org/opencypher/caps/impl/spark/SparkSQLExprMapper.scala]: Duplicate doc row

File: caps-spark/src/main/scala/org/opencypher/caps/impl/spark/physical/operators/BinaryOperators.scala [hunk 8]
```diff
@@ -53,10 +53,20 @@ final case class ValueJoin(
   }
 
 }
-abstract class PatternJoin extends BinaryPhysicalOperator {
+
+/**
+  * This operator performs a left outer join between the already matched path and the optional matched pattern and
+  * updates the resulting columns.
+  *
+  * @param lhs mandatory match data
```
Comment (Mats-SX) [inline caps-spark/src/main/scala/org/opencypher/caps/impl/spark/physical/operators/BinaryOperators.scala]: let's not use `mandatory`; I prefer `non-optional` or `matched` or `definitely matched`

File: caps-spark/src/main/scala/org/opencypher/caps/impl/spark/physical/operators/BinaryOperators.scala [hunk 9]
```diff
@@ -53,10 +53,20 @@ final case class ValueJoin(
...
+  * @param rhs optional match data
+  * @param header result header (lhs header + rhs header)
+  */
+final case class Optional(lhs: PhysicalOperator, rhs: PhysicalOperator, header: RecordHeader)
+    extends BinaryPhysicalOperator {
 
   override def executeBinary(left: PhysicalResult, right: PhysicalResult)
-                            (implicit context: RuntimeContext): PhysicalResult = {
+      (implicit context: RuntimeContext): PhysicalResult = {
```
Comment (Mats-SX) [inline caps-spark/src/main/scala/org/opencypher/caps/impl/spark/physical/operators/BinaryOperators.scala]: strange indent change

File: caps-core/src/main/scala/org/opencypher/caps/api/expr/Expr.scala (hunk 1 above)
Comment (DarthMax) [reply caps-core/src/main/scala/org/opencypher/caps/api/expr/Expr.scala]: Yes, so we only do this for this single expression which clearly signals: This is a subquery, do plan it.

File: caps-core/src/main/scala/org/opencypher/caps/impl/logical/LogicalPlanner.scala (hunk 3 above)
Comment (DarthMax) [reply caps-core/src/main/scala/org/opencypher/caps/impl/logical/LogicalPlanner.scala]: It does, but the semantic check do not allow it I guess

File: caps-core/src/test/scala/org/opencypher/caps/impl/typer/SchemaTyperTest.scala (hunk 6 above)
Comment (DarthMax) [reply caps-core/src/test/scala/org/opencypher/caps/impl/typer/SchemaTyperTest.scala]: because the test helper does not call rewriters and sub queries can not yet be parsed. We building the expression tree by hand but it was a mess and would not have contributed much in terms of stability etc

File: caps-core/src/main/scala/org/opencypher/caps/impl/logical/LogicalPlanner.scala (hunk 3 above)
Comment (Mats-SX) [reply caps-core/src/main/scala/org/opencypher/caps/impl/logical/LogicalPlanner.scala]: It does, but that's another debate really. (Because in Cypher 3.x pattern expressions evaluate to a list of paths and can be used in any expression context.)

File: caps-spark/src/main/scala/org/opencypher/caps/impl/spark/physical/operators/BinaryOperators.scala (hunk 9 above)
Comment (DarthMax) [reply caps-spark/src/main/scala/org/opencypher/caps/impl/spark/physical/operators/BinaryOperators.scala]: mhh I don't find a better way tbh

File: caps-core/src/main/scala/org/opencypher/caps/impl/logical/LogicalPlanner.scala (hunk 3 above)
Comment (DarthMax) [reply caps-core/src/main/scala/org/opencypher/caps/impl/logical/LogicalPlanner.scala]: mhh I guess that is a point. On the other hand we do not really have a good way to restrict this in the planInnerExpression method right now. We could introduce special cases for `NOT`, `ANDS` and `ORS` but that does not feel good

File: caps-core/src/main/scala/org/opencypher/caps/impl/util/MapUtils.scala (hunk 5 above)
Comment (s1ck) [reply caps-core/src/main/scala/org/opencypher/caps/impl/util/MapUtils.scala]: You're right. We weren't away of that.

Comment (s1ck) [reply caps-core/src/main/scala/org/opencypher/caps/impl/util/MapUtils.scala]: However, I think we should keep the method as it's much more readable.

File: caps-core/src/main/scala/org/opencypher/caps/impl/logical/LogicalPlanner.scala (hunk 3 above)
Comment (Mats-SX) [reply caps-core/src/main/scala/org/opencypher/caps/impl/logical/LogicalPlanner.scala]: I agree.
